mitodl/mit-open
# Update React Router to v6

## What are the relevant tickets?
#238 

## Description (What does it do?)
This PR updates our react router version from v5 to v6. There are some significant differences.

### Most notable changes:
- **No more `history` object.**  React router has always (and still) uses another package, `history` to implement its history stack for browser navigation. In `react-router` v5 and below, the `history` package was part of its public API. Now it's an implementation detail. As such, we can't access `history` directly.
    - Most usage of `history.push` (or `history.replace`) is replaceable by a hook, `useNavigate`. 
    - A significant exception to this is our API error handling code. See [this comment](https://github.com/mitodl/mit-open/pull/250/files#r1409864751) for details.
- Routes now defined as an array of objects (Rather than JSX) in `routes.tsx` 

## How can this be tested?
After `docker compose up`ing on this branch...

As an authenticated admin user:
1. Visit http://localhost:8063/learningpaths ... Navigate to individual paths. Things should work.
2. Visit http://localhost:8063/articles/new ... and create an article. Then edit it and save. Routing should work appropriately.
3. Test Refactored 404: Visit http://localhost:8063/articles/not_exist_id
    - You will se a development-only overlay. It's annoying. Click the X in upper right corner. (Possibly twice.)
    -  Then you'll see the 404 page.

Test the Refactered 403 handling:
1. In a private browser, view http://localhost:8063/articles/SOME_ARTICLE_ID. You'll be redirected to a login page. 
2.Logged in as a user without superuser/admin privileges, visit http://localhost:8063/articles/SOME_ARTICLE_ID. You should see a 403 page.
